nad_ch/controllers/cli.py

```diff
@@ -11,28 +15,33 @@ def cli(ctx):
 
 @cli.command()
 @click.pass_context
-@click.argument('filename')
-@click.argument('content')
-def upload(ctx, filename, content):
+@click.argument('provider_name')
+def add_provider(ctx, provider_name):
     context = ctx.obj
-    file = File(name=filename, content=content)
-    upload_file(context, file)
-    click.echo(f"Uploaded {filename}")
+    try:
+        add_data_provider(context, provider_name)
+    except InvalidProviderNameException as e:
+        click.echo(f"Error: {e.message}")
+        return
```

Comment:
Similar here to below - it's debatable, but I'd consider putting the exception handling and logging in the use case. That way the controller just needs to know what command to invoke.

Reply:
Unless I'm misunderstanding, wouldn't that limit the utility of the use case? I.e., an HTTP controller of some sort wouldn't be able to use it since it's logging directly?

Might another way of splitting the presentation logic out of the CLI controller be to make some sort of CLI presenter that handles the logging? Such that the presenter accepts the app context and then uses the logging interface to do whatever it needs to do.

Reply:
IMO if the code can be written to be more abstract and not needlessly complicate matters, it's better to not couple it to the CLI interface. Since it's a command-like thing, where the result is a side effect (adding something to the database), the caller only needs to know if it succeeded or failed. Logging an info message is a bonus, which you could continue with an HTTP route also. If logging were a requirement, you could have test coverage over the more abstract use case, rather than the more volatile controller. If you needed to return an error message via an HTTP API, you could pass it as a return value, while maintaining the logging, or like you suggested, inject some more specific interface like a presenter.

Anyway, it's pretty minor as it's just one line of logic, and the concerns applies more if the error handling was more complex, so it's probably a pedantic concern.

Reply:
I *think* I understand what you're driving at. I do think it makes sense to sort this out as early as possible since other use cases will probably have more complex error handling requirements. [Here's](https://github.com/GSA-TTS/10x-nad-ch/pull/17/commits/cbd6004a6f3ff0acb5c1ef35aeaf8a04c56d7696) an attempt to address this.